### PR TITLE
Fix keybindings and update to 2018 edition. Closes #11.

### LINKED ISSUE #11: keybinds not working
Howdy!

Neat project. Ive been giving it a go and have run into a snag with the custom keybinds.

Ive set the env var and pointed it to the default.keys found in this repo. It errors out with the below until I remove he lines for `undo_delete` and `help`:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: Other, error: StringError("invalid config at line 30: undo_delete:C-z") }) }', src/libcore/result.rs:860
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Removing those gets it to launch, but the app still responds to the default bindings and does not appear to notice any changes I make.

Have you run into this before?

### PR DESCRIPTION
**Pull request by alok**
This lets us map keys such as `A-x`, `A-?` (alt + shift + /), ':', and '?',
which closes #11.

I ended up making a bunch of more cosmetic changes in the process, and was too
lazy to split them off into separate PRs.

### CODE REVIEW

**Comment by alok** [comment]
@spacekookie this one is more up to date than #23

File: default.keys [hunk 1]
```diff
@@ -1,3 +1,8 @@
+# This is a comment
+# Blank lines are also supported
+# You can put spaces between the colons
```
Comment (spacekookie) [inline default.keys]: I feel like this description doesn't need to be in this file. It's kinda obvious or we should document it elsewhere

File: src/config.rs [hunk 2]
```diff
@@ -178,30 +179,38 @@ impl fmt::Display for Config {
...
+            }
+
+            // zero based indexing inappropriate here
+            line_num += 1;
+            // TODO Separate error messages to be more informative in each of the two cases
+            // below.
+            let e = format!("invalid config at line {}: {}", line_num, line);
 
-            let parts: Vec<_> = line.split(":").map(|p| p.trim()).collect();
+            let parts: Vec<_> = line.splitn(2, ':').map(|p| p.trim()).collect();
```
Comment (spacekookie) [inline src/config.rs]: What's the advantage of this?

File: src/screen.rs [hunk 3]
```diff
@@ -563,7 +562,7 @@ impl Screen {
                     error!("command failed to start: {}", content);
                 }
             } else {
-                let shell = env::var("SHELL").unwrap_or("bash".to_owned());
+                let shell = env::var("SHELL").unwrap_or_else(|_| "bash".to_owned());
```
Comment (spacekookie) [inline src/screen.rs]: Is this a clippy lint? If so, okay.
If not, I'm not sure if it's really quite worth the optimisation

File: src/bin/void.rs [hunk 4]
```diff
@@ -1,17 +1,15 @@
-extern crate getopts;
-extern crate fs2;
 extern crate dirs;
+extern crate fs2;
+extern crate getopts;
 extern crate voidmap;
```
Comment (spacejam) [inline src/bin/void.rs]: feel free to drop all `extern crate`'s

File: .travis.yml [hunk 5]
```diff
@@ -1,12 +1,8 @@
 language: rust
```
Comment (spacekookie) [inline .travis.yml]:
```suggestion
language: rust
rust:
  - stable
  - nightly
  - nightly-2019-02-02
script: |
  cargo build --verbose
  cargo test --verbose --all
matrix:
  include:
    - rust: nightly-2019-02-02
      env: CLIPPY=YESPLEASE
      before_script: |
        rustup component add clippy
        rustup component add rustfmt
      script: |
        cargo clippy --all -- -D warnings
        cargo fmt --all -- --check
```

File: src/config.rs [hunk 6]
```diff
@@ -167,9 +167,10 @@ impl Default for Config {
 
 impl fmt::Display for Config {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "Configured Hotkeys:\n").unwrap();
+        writeln!(f, "Configured Hotkeys:").unwrap();
         for (key, action) in &self.config {
-            write!(f, "    {:?}: {:?}\n", action, key).unwrap();
+            // writeln!(f, "    {:?}: {:?}", action, key).unwrap();
```
Comment (spacejam) [inline src/config.rs]: please remove commented-out code

File: src/config.rs [hunk 7]
```diff
@@ -86,32 +85,33 @@ fn str_to_action(input: String) -> Option<Action> {
...
-                .nth(0)
-                .and_then(|n| n.at(1))
-                .and_then(|r| r.chars().nth(0))
-                .map(|c| Ctrl(c))
-        }
+
+        key if key.len() == 1 => Some(Char(extract_key(key, 0))),
+
+        key if key.starts_with("A-") => Some(Alt(extract_key(key, 2))),
+        key if key.starts_with("C-") => Some(Ctrl(extract_key(key, 2))),
```
Comment (spacejam) [inline src/config.rs]: when `extract_key` returns an `Option`, no need to wrap in `Some`, but instead, `extract_key(key, ...).map(|Ctrl|)` etc...

**Comment by spacekookie** [comment]
Also, I just ran `rustfmt 1.0.0` (on `1.32` stable) on this PR and it changed a lot of code. Am I missing some config to configure it or are we running two different versions?

File: src/config.rs [hunk 8]
```diff
@@ -178,30 +179,38 @@ impl fmt::Display for Config {
...
-            let e = format!("invalid config at line {}: {}", line_number, line);
+        for (mut line_num, line) in buf.lines().enumerate() {
+            if line == "" || line.starts_with('#') {
+                continue;
+            }
+
+            // zero based indexing inappropriate here
+            line_num += 1;
+            // TODO Separate error messages to be more informative in each of the two cases
+            // below.
```
Comment (spacejam) [inline src/config.rs]: please either implement the item in the `TODO` or open an issue that refers to this function. I'd prefer not to commit `TODO`'s

File: default.keys (hunk 1 above)
Comment (alok) [reply default.keys]: I plan on doing a follow up PR that uses something like TOML to do config, with something like a `[keys]` section, so this is more a stopgap.

File: src/config.rs (hunk 2 above)
Comment (alok) [reply src/config.rs]: Allows commands like `A-:` to be mapped.

File: src/screen.rs (hunk 3 above)
Comment (alok) [reply src/screen.rs]: clippy

File: src/lib.rs [hunk 9]
```diff
@@ -1,49 +1,48 @@
...
 extern crate hyper;
+extern crate libc;
+extern crate protobuf;
 extern crate rand;
+extern crate regex;
 extern crate termion;
-extern crate protobuf;
-extern crate libc;
+extern crate time;
+extern crate unicode_segmentation;
```
Comment (spacejam) [inline src/lib.rs]: feel free to remove all `extern crate`'s

File: test/test.rs [hunk 10]
```diff
@@ -1,6 +1,6 @@
 extern crate libc;
-extern crate voidmap;
-extern crate termion;
 extern crate quickcheck;
 extern crate rand;
+extern crate termion;
+extern crate voidmap;
```
Comment (spacejam) [inline test/test.rs]: feel free to remove all `extern crate`'s

File: src/screen.rs [hunk 11]
```diff
@@ -115,9 +122,9 @@ impl Default for Screen {
 impl Screen {
     fn help(&mut self) {
         self.cleanup();
-        print!("{}{}{}\n", cursor::Goto(1, 1), clear::All, self.config);
+        println!("{}{}{}", cursor::Goto(1, 1), clear::All, self.config);
```
Comment (spacejam) [inline src/screen.rs]: `println` acquires a Mutex and writes the line synchronously, but `print` allows things to buffer up more before flushing (at least last time I looked into this). If that's no longer the case, then this is fine, but please double check and keep it as a `print` if the lock is acquired.

Comment (alok) [reply src/screen.rs]: I just looked at the source code for both and they use the same function (`io::_print`) to do it, except println uses a macro that adds a newline at the end.

**Comment by spacekookie** [comment]
`cargo fmt --all` is still producing a lot of code changes for me. Maybe have a look at what version you're on?

**Comment by alok** [comment]
It should be changing code, since I didn't run it yet to keep the diff clean.

**Comment by spacekookie** [comment]
:+1: Maybe just run that at the end then and we can squash all the changes. Or we can do it after merging

**Comment by alok** [comment]
After merge is probably easier.

**Comment by spacekookie** [comment]
Alright, thank you so much for this PR! :heart: 
I'm gonna merge it now and clean up the rest (i.e. fix CI and formatting)

**Pull request merged**